woocommerce/woocommerce-gutenberg-products-block
# Customize your Store > Implement retry mechanism, enhance prompts and fix issue with AI content not being generated for new products

## What

On this PR, a series of changes are made to ensure that:

- Whenever AI fails to generate content for Patterns, Products, or while managing images, a retry mechanism is triggered up to five times, with an error being triggered only after the fifth failed attempt.
- Prompts to AI are enhanced for products, patterns, and images.
- The method responsible for fetching the collection of dummy products to be updated by AI `fetch_dummy_products_to_update` is not invoked on the `product` endpoint but on the `products` endpoint instead. This solves the regression where this method would be invoked multiple times (for each product), resulting in too many products being programmatically updated on each request and not all of them having AI-managed/generated content assigned to them.
- Ensure the action for updating the business description is scheduled only if the provided description differs from the previously saved one.

The following diagrams demonstrate the updated flow with the changes made on this PR in place:

### AI images management:

![image_flow_2](https://github.com/woocommerce/woocommerce-blocks/assets/15730971/9976a3a7-2af7-4c7a-98f2-4815309216ee)

### AI content generation for products and patterns:

![content_flow](https://github.com/woocommerce/woocommerce-blocks/assets/15730971/a6eb7977-6b79-4ddc-b5a4-bf0268fb1460)

## Why



These changes ensure AI content generation and image management are more consistent and reliable.

## Testing Instructions



_Please consider any edge cases this change may have, and also other areas of the product this may impact._

### Setup

Please test the feature on a WooExpress install instead of Jurassic Ninja, as the latter can be very limited in terms of server resources and unstable, as other installs can interfere with your own.

- If you already have a pre-configured install, remove the pre-existing WooCommerce Blocks Plugin and upload the one from this zip file: [woocommerce-gutenberg-products-block.zip](https://github.com/woocommerce/woocommerce-blocks/files/13415580/woocommerce-gutenberg-products-block.zip)
- Alternatively, create a new WooExpress install from scratch 
- Remove the pre-existing WooCommerce plugin via SSH and
- Install and activate WooCommerce from the following zip file: 
[woocommerce.zip](https://github.com/woocommerce/woocommerce-blocks/files/13415324/woocommerce.zip)
- Install and activate WooCommerce Blocks from the following zip file: [woocommerce-gutenberg-products-block.zip](https://github.com/woocommerce/woocommerce-blocks/files/13415580/woocommerce-gutenberg-products-block.zip)
- Install and activate WooCommerce Beta Tester (the plugin is available [within the monorepo](https://github.com/woocommerce/woocommerce/tree/trunk/plugins))
- Head over to `/wp-admin/tools.php?page=woocommerce-admin-test-helper` and enable `customize-store` feature flag:

<img width="1233" alt="Screenshot 2023-10-22 at 10 32 24" src="https://github.com/woocommerce/woocommerce-blocks/assets/15730971/816867c6-34e5-40a1-a87d-f4a7cda46429">

### Test the CYS experience

0. Make sure you don't have any products in your store.
1. Head over to Home > Customize your Store:
 
<img width="1232" alt="Screenshot 2023-10-22 at 09 29 08" src="https://github.com/woocommerce/woocommerce-blocks/assets/15730971/b5d34131-3edc-47ed-a8e4-8e2775ca1bbd">

2. Click on design with AI.
3. Type a description for your business.
4. Make sure six new dummy products are created for your store.
5. Make sure the AI content is properly generated and images are assigned to patterns and products based on the description provided.

Caveat: the AI endpoint is known for being unstable, so you might encounter errors during the loading screen. If that happens, click "Try Again" to attempt updating one more time. We have additional upcoming changes to ensure better error handling planned for implementation in upcoming PRs.


## Screenshots or screencast




https://github.com/woocommerce/woocommerce-blocks/assets/15730971/2ccecc54-d1a5-4517-b190-9697ff02a6c6



## WooCommerce Visibility


Required:


## Checklist

Required:


Conditional:
